src/VisualStudio/Core/Def/Implementation/IntellisenseControls/IntellisenseTextBox.cs

```diff
@@ -0,0 +1,528 @@
+                {
+                    ITextSnapshot textSnapshot = this._textViewHost.TextView.TextSnapshot;
+                    if (this.Text.IsNullOrWhiteSpace())
+                    {
+                        // Select all text in the control if edit is not started by user keyboard input
+                        ITextSelection textSelection = this._textViewHost.TextView.Selection;
+                        textSelection.Select(new SnapshotSpan(textSnapshot, 0, textSnapshot.Length), false);
+                    }
+                    else
+                    {
+                        // Otherwise Move to caret to end
+                        ITextCaret textCaret = this._textViewHost.TextView.Caret;
+                        textCaret.MoveTo(new SnapshotPoint(textSnapshot, textSnapshot.Length));
+                    }
+                }
+            }
+        }
+
+        /// <summary>
+        /// Determines if we can cut or copy in the editor control
+        /// </summary>
+        /// <returns>True if cut and copy operation should be enabled</returns>
+        private bool CanCutCopy()
+        {
+            if (this._textViewHost != null)
+            {
+                ITextSelection textSelection = this._textViewHost.TextView.Selection;
+
+                if (textSelection.SelectedSpans != null && textSelection.SelectedSpans.Count > 0)
+                {
+                    foreach (SnapshotSpan snapshotSpan in textSelection.SelectedSpans)
+                    {
+                        if (snapshotSpan.Length > 0)
+                        {
+                            return true;
+                        }
+                    }
+                }
+            }
+
+            return false;
+        }
+
+        [DllImport("user32.dll")]
+        [PreserveSig]
+        private static extern int GetPriorityClipboardFormat([In] uint[] formatPriorityList, [In] int formatCount);
+
+        /// <summary>
+        /// Determines if we can paste to the editor control
+        /// </summary>
+        /// <returns>True, if paste is supported</returns>
+        private bool CanPaste()
+        {
+            const uint TextFormatId = 1;          // ID for clipboard text format (CF_TEXT)
+            const uint UnicodeTextFormatId = 13;  // ID for clipboard unicode text format (CF_UNICODETEXT)
+
+            // We should avoid calling Clipboard.GetDataObject() as it will hang if the clipboard owner is blocked while
+            // putting data into the clipboard using delay rendering (Dev14 bug: 1142153, Dev10 bug: 788188)
+            uint[] formats = new uint[] { TextFormatId, UnicodeTextFormatId };
+
+            // GetPriorityClipboardFormat returns 0 if the clipboard is empty, -1 if there is data in the clipboard but it does not match
+            // any of the input formats and otherwise returns the ID of the format available in the clipboard.
+            return GetPriorityClipboardFormat(formats, formats.Length) > 0;
+        }
```

Comment:
Seems like this should live in the editor and not us?

Reply:
@amadeusw, @tmat, @dustincoleman, @jinu
We definitely do not want to keep this code in Roslyn. 
This code (and supplementary classes) can be used by Roslyn, Debugger and Interactive window. Maybe by some other teams as well. It seems generic and not Roslyn-specific. 

Please consider moving it to the Editor platform.


---
In reply to: [367169980](https://github.com/dotnet/roslyn/pull/38748#discussion_r367169980) [](ancestors = 367169980)